agimus_controller_ros/agimus_controller_ros/agimus_controller.py

```diff
@@ -0,0 +1,250 @@
+        horizon_points = self.traj_buffer.get_points(
+            self.params.ocp.horizon_size, self.point_attributes
+        )
+        x_plan = np.zeros([self.params.ocp.horizon_size, self.nx])
+        a_plan = np.zeros([self.params.ocp.horizon_size, self.nv])
+        for idx_point, point in enumerate(horizon_points):
+            x_plan[idx_point, :] = point.get_x_as_q_v()
+            a_plan[idx_point, :] = point.a
+
+        # First solve
+        self.ocp = OCPCrocoHPP(self.rmodel, self.cmodel, self.params.ocp)
+        self.ocp.set_planning_variables(x_plan, a_plan)
+        self.mpc = MPC(self.ocp, x_plan, a_plan, self.rmodel, self.cmodel)
+        us_init = self.mpc.ocp.u_plan[: self.params.ocp.horizon_size - 1]
+        self.mpc.mpc_first_step(x_plan, us_init, x0)
+        self.next_node_idx = self.params.ocp.horizon_size
+        if self.params.save_predictions_and_refs:
+            self.mpc.create_mpc_data()
+        _, u, k = self.mpc.get_mpc_output()
+        return u, k
+
+    def solve(self, x0):
+        self.target_translation_object_to_effector = None
+        self.fill_buffer()
+        if self.traj_buffer.get_size(self.point_attributes) > 0:
+            point = self.traj_buffer.get_points(1, self.point_attributes)[0]
+            self.last_point = point
+            new_x_ref = point.get_x_as_q_v()
+        else:
+            point = self.last_point
+            new_x_ref = point.get_x_as_q_v()
+        new_a_ref = point.a
+        self.in_world_M_effector = get_ee_pose_from_configuration(
+            self.rmodel,
+            self.rdata,
+            self.effector_frame_id,
+            new_x_ref[: self.rmodel.nq],
+        )
+        self.mpc.mpc_step(x0, new_x_ref, new_a_ref, self.in_world_M_effector)
+
+        if self.params.save_predictions_and_refs:
+            self.mpc.fill_predictions_and_refs_arrays()
+        _, u, k = self.mpc.get_mpc_output()
+        return u, k
+
+    def get_sensor_msg(self):
+        sensor_msg = deepcopy(self.sensor_msg)
+        return sensor_msg
+
+    def send(self, sensor_msg, u, k):
+        self.control_msg.header = Header()
+        self.control_msg.header.stamp = Time.time()
+        self.control_msg.feedback_gain = matrix_numpy_to_msg(k)
+        self.control_msg.feedforward = matrix_numpy_to_msg(u)
+        self.control_msg.initial_state = sensor_msg
+        self.control_publisher.publish(self.control_msg)
+
+    def exit_handler(self):
+        print("saving data")
+        np.save("mpc_params.npy", self.params.get_dict())
+        if self.params.save_predictions_and_refs:
+            np.save("mpc_data.npy", self.mpc.mpc_data)
+
+    def get_x0_from_sensor_msg(self, sensor_msg):
+        return np.concatenate(
+            [sensor_msg.joint_state.position, sensor_msg.joint_state.velocity]
+        )
+
+    def publish_ocp_solve_time(self, ocp_solve_time):
+        self.ocp_solve_time_pub.publish(
+            convert_float_to_ros_duration_msg(ocp_solve_time)
+        )
```

Comment:
This might be a part of utils

Reply:
you think `publish_ocp_solve_time` function should be moved to something like `agimus_controller_ros/sim_utils.py ` ?